docs/guides/high-availability/concepts-memory-and-cpu-sizing.adoc

```diff
@@ -0,0 +1,95 @@
+<#import "/templates/guide.adoc" as tmpl>
+<#import "/templates/links.adoc" as links>
+
+<@tmpl.guide
+title="Concepts for sizing CPU and memory resources"
+summary="Understand these concepts to avoid resource exhaustion and congestion"
+preview="true" >
+
+Use this as a starting point to size a product environment. 
+Adjust the values for your environment as needed based on your load tests.
+
+== Performance recommendations
+
+[WARNING]
+====
+* Performance will be lowered when scaling to more Pods (due to additional overhead) and using a cross-datacenter setup (due to additional traffic and operations).
+
+* Increased cache sizes can improve the performance when {project_name} instances run for a longer time. Still, those caches need to be filled when an instance is restarted.
+
+* Use these values as a starting point and perform your own load tests before going into production.
+====
+
+Summary:
+
+* The used CPU scales linearly with the number of requests up to the tested limit below.
+* The used memory scales linearly with the number of active sessions up to the tested limit below.
+
+Recommendations:
+
+* The base memory usage for an inactive Pod is 1 GB of RAM.
+
+* Leave 1 GB extra head-room for spikes of RAM.
+
+* For each 100,000 active user sessions, add 500 MB per Pod in a three-node cluster (tested with up to 200,000 sessions).
++
+This assumes that each user connects to only one client.
+Memory requirements increase with the number of client sessions per user session (not tested yet).
+
+* For each 40 user logins per second, 1 vCPU per Pod in a three-node cluster (tested with up to 300 per second).
++
+{project_name} spends most of the CPU time hashing the password provided by the user.
+
+* For each 450 client credential grants per second, 1 vCPU per Pod in a three node cluster (tested with up to 2000 per second).
++
+Most CPU time goes into creating new TLS connections, as each client runs only a single request.
+
+* For each 350 refresh token requests per second, 1 vCPU per Pod in a three node cluster (tested with up to 435 refresh token requests per second).
+
+* Leave 200% extra head-room for CPU usage to handle spikes in the load.
+This ensures a fast startup of the node, and sufficient capacity to handle failover tasks like, for example, re-balancing Infinispan caches, when one node fails.
+Performance of {project_name} dropped significantly when its Pods were throttled in our tests.
+
+=== Calculation example
```

Comment:
In my experience, the numbers used in the calculation show a low to medium utilization scenario.
This is probably the scenario that most (~80%) Keycloak users will encounter.

However, 10-20% of the Keycloak projects I have seen have a high to very high utilization scenario, e.g. 1,000,000 - 10,000,000 concurrent user sessions (with at least one client).

In the previous explanation it is written that CPU and memory scale linearly. Nevertheless, I suggest to explicitly explain a high load scenario with 5,000,000 user sessions.

Reply:
In addition, the user sessions of users do not have to remain in RAM forever.
In the case of offline user/client sessions, the memory requirement can also be reduced somewhat over time by shortening the cache entry lifetime. 
https://github.com/keycloak/keycloak/discussions/24274

Reply:
@thomasdarimont for your two comments, please create two new issues in the KCB project for this: https://github.com/keycloak/keycloak-benchmark/issues

We could over there collaborate on this and track the progress.